### PR TITLE
Fix config file location

### PR DESCRIPTION
**Pull request by kang-makes**
Solves #313

### CODE REVIEW

File: internal/config/config.go [hunk 1]
```diff
@@ -4,12 +4,13 @@ import (
 	"strings"
 	"time"
 
+	"github.com/spf13/pflag"
 	"github.com/spf13/viper"
 )
 
 const (
-	FileName = "nri-kubernetes-config"
-	FilePath = "/etc/newrelic-infra/integrations.d/"
```
Comment (paologallinaharbur) [inline internal/config/config.go:12]: I read the issue but I did not understand why the location is invalid and what you mean by "That path is for the agent, not the integration itself."
The charts looks like it is mounting the file exacting there and so far always worked.

File: internal/config/config.go [hunk 2]
```diff
@@ -99,9 +125,10 @@ func LoadConfig(filePath string, fileName string) (*Config, error) {
 
 	// Config File
 	v.AddConfigPath(filePath)
+	v.AddConfigPath(".")
```
Comment (paologallinaharbur) [inline internal/config/config.go:128]: This is useful when running it locally, we were missing it 👍

Comment (paologallinaharbur) [reply internal/config/config.go:128]: Now that I think about it.. should we create a const for it?

File: internal/config/config_test.go [hunk 3]
```diff
@@ -18,15 +18,15 @@ func TestLoadConfig(t *testing.T) {
...
+		c, err := config.LoadConfigWithFileLocation(fakeDataDir, workingData)
 		require.NoError(t, err)
 		require.Equal(t, "dummy_cluster", c.ClusterName)
 
 		t.Run("with_env_precedence", func(t *testing.T) {
 			_ = os.Setenv("NRI_KUBERNETES_CLUSTERNAME", "different_value")
 			_ = os.Setenv("NRI_KUBERNETES_NODENAME", "fake-node")
 
-			c, err := config.LoadConfig(fakeDataDir, workingData)
+			c, err := config.LoadConfigWithFileLocation(fakeDataDir, workingData)
```
Comment (paologallinaharbur) [inline internal/config/config_test.go:29]: I would add a test for loadConfig as well

File: internal/config/config.go [hunk 4]
```diff
@@ -81,7 +82,32 @@ type MTLS struct {
...
-func LoadConfig(filePath string, fileName string) (*Config, error) {
+func LoadConfig() (*Config, error) {
+	v := viper.New()
+
+	pflag.StringP("config-name", "c" , FileName, "filename of the configuration to look for")
+	pflag.StringP("config-path", "p" , FilePath, "path where the configuration file is")
+
+	// Environment variables instead of v consistent with the config file
+	v.SetEnvPrefix("NRI_KUBERNETES")
+	v.AutomaticEnv()
```
Comment (paologallinaharbur) [inline internal/config/config.go:93]: Do we want to support env for these flags?

File: internal/config/config.go [hunk 5]
```diff
@@ -81,7 +82,32 @@ type MTLS struct {
 	TLSSecretNamespace string `mapstructure:"secretNamespace"`
 }
 
-func LoadConfig(filePath string, fileName string) (*Config, error) {
+func LoadConfig() (*Config, error) {
+	v := viper.New()
```
Comment (paologallinaharbur) [inline internal/config/config.go:86]: Can't we leverage the same viper "obj" or is it cleaner in this way?

File: internal/config/config.go [hunk 6]
```diff
@@ -99,9 +125,10 @@ func LoadConfig(filePath string, fileName string) (*Config, error) {
 
 	// Config File
 	v.AddConfigPath(filePath)
+	v.AddConfigPath(".")
 	v.SetConfigName(fileName)
 
-	// If error reading file or file not found, use flag/env variables
```
Comment (paologallinaharbur) [inline internal/config/config.go:104]: I do not understand neither the old or the new comment 😅

File: internal/config/config.go (hunk 1 above)
Comment (kang-makes) [reply internal/config/config.go:12]: I made a PR for the helm chart as well: https://github.com/newrelic/helm-charts/pull/638

`/etc/newrelic-infra/integrations.d/` is the path were the agent read the configurations for integrations, not the config for the integration itself.

As the agent is if forward only mode there is so much issue with that but in the case of kubelet daemonset, the agent compalins of that  file not being a valid config v3/v4 config file.

TL;DR: Agent expect [this](https://docs.newrelic.com/docs/infrastructure/host-integrations/infrastructure-integrations-sdk/specifications/host-integrations-standard-configuration-format/#config) instead of the configuration for the integration itself.

File: internal/config/config.go (hunk 2 above)
Comment (kang-makes) [reply internal/config/config.go:128]: I thought about it but then it would be an array/slice and slices are not constants in Go.

I accept suggestions.

File: internal/config/config.go (hunk 4 above)
Comment (kang-makes) [reply internal/config/config.go:93]: If could make the development easier or could change the location using envs instead of flags.

I would support everything just in case. it could be a NTH in the future to use remote configuration, for example.

File: internal/config/config.go (hunk 1 above)
Comment (paologallinaharbur) [reply internal/config/config.go:12]: The chart should not mount that as well in the agent sidecar, therefore  the agent should not aware of this at all.

However, it is true what you say and `/etc/newrelic-infra` looks better

File: internal/config/config.go (hunk 5 above)
Comment (kang-makes) [reply internal/config/config.go:86]: If I use the same Viper object (and thus I can reuse the same function), Viper will try to look for `configPath`(or was it `config-path`?) and `configName` in the YAML and fails because they are not optional yada yada...

**Comment by paologallinaharbur** [comment]
Related to https://github.com/newrelic/helm-charts/pull/638/files

File: internal/config/config.go (hunk 1 above)
Comment (kang-makes) [reply internal/config/config.go:12]: I am rookie so I could be wrong. Correct me if I am wrong, please.

The agent does not need of this file because is a configuration for the integration itself, the agent should not have any configuration file for this integration as it should not run/execute any binary for this integration.

What the agent should have is an open port were this integration will POST the metrics that the agent should send to the backend.

AFAIK the agent do not need anything for that except forwardMode and forwardOnlyMode, right?

File: internal/config/config.go (hunk 6 above)
Comment (kang-makes) [reply internal/config/config.go:104]: Fixed in a nit picking commit. Would you resolve this conversation?

File: internal/config/config.go (hunk 1 above)
Comment (paologallinaharbur) [reply internal/config/config.go:12]: everything correct 😄 
The agent container should not be even aware of this file, therefore you were right that `integrations.d/` looks weird

File: internal/config/config.go [hunk 7]
```diff
@@ -81,10 +82,35 @@ type MTLS struct {
 	TLSSecretNamespace string `mapstructure:"secretNamespace"`
 }
 
-func LoadConfig(filePath string, fileName string) (*Config, error) {
+func LoadConfig() (*Config, error) {
 	v := viper.New()
 
-	// We need to assure that defaults have been set in order to bind env variables.
+	pflag.StringP("config-name", "c", FileName, "filename of the configuration to look for")
+	pflag.StringP("config-path", "p", FilePath, "path where the configuration file is")
```
Comment (roobre) [inline internal/config/config.go:89]: These seem kind of redundant to me. I'd expect to either look for a hardcoded name (`nri-kubernetes.yaml`) in a harcoded list of paths (`.`, `/etc/newrelic-infra`), or to just point to the full path directly.

File: internal/config/config.go [hunk 8]
```diff
@@ -81,10 +82,35 @@ type MTLS struct {
...
+
+	// If error parsing flag/env variables
+	if err := v.BindPFlags(pflag.CommandLine); err != nil {
+		return nil, err
+	}
+
+	return LoadConfigWithFileLocation(
+		v.GetString("config-path"),
+		v.GetString("config-name"),
+	)
```
Comment (roobre) [inline internal/config/config.go:107]: Could we get rid of this if we use fullpath-or-default as suggested above?

File: internal/config/config.go (hunk 2 above)
Comment (kang-makes) [reply internal/config/config.go:128]: Done.

File: internal/config/config_test.go (hunk 3 above)
Comment (kang-makes) [reply internal/config/config_test.go:29]: Done.

File: internal/config/config.go (hunk 2 above)
Comment (paologallinaharbur) [reply internal/config/config.go:128]: I mean simply to add a separate constant like ` const localPath = "."` instead of an inline variable.

The other approach would be to use a func just for that:
```golang
func getSlice() []string {
    return []string{"hello", "world"}
}
```
but maybe it is too much

File: internal/config/config.go (hunk 7 above)
Comment (paologallinaharbur) [reply internal/config/config.go:89]: It makes sense to have them divided since Viper is asking for those separately and therefore we would need to split it, no?

**Review by paologallinaharbur — APPROVED** [review]
Overall it looks ok!

Pre-approving, but there is still one nit open and a conversation from Roberto

**Comment by kang-makes** [comment]
In the end, the path with `integrations.d` was not making any issues because the configuration was mounted in the integration container but not in the agent sidecar but this can lead to misunderstandings.

I have ditched all the work from the last 2 hours because tests were passing but when I tried to run the integration itself failed because the Integration SDK also manages flags its own way.

```
flag provided but not defined: -c
Usage of /private/var/folders/n0/x5b9pcyx5f3528l599g3mymh0000gn/T/GoLand/___2go_build_github_com_newrelic_nri_kubernetes:
Exiting.
```

So I simply added the local folder as a source were to load the configuration file and changed path where to load the integration configuration.

**Review by paologallinaharbur — APPROVED** [review]
LGTM (😢 SDK)

File: internal/config/config.go [hunk 9]
```diff
@@ -84,7 +84,7 @@ type MTLS struct {
 func LoadConfig(filePath string, fileName string) (*Config, error) {
 	v := viper.New()
 
-	// We need to assure that defaults have been set in order to bind env variables.
+	// We need to assure that defaults have been set in order to bind env variables
```
Comment (roobre) [inline internal/config/config.go:87]: Bring the trailing dot back 😠

**Review by roobre — APPROVED** [review]

File: internal/config/config.go (hunk 9 above)
Comment (kang-makes) [reply internal/config/config.go:87]: https://github.com/newrelic/newrelic-infra-operator/blob/ce13dfff70008c54ed0598bb6904b0cf8fed9808/.golangci.yml#L66

https://github.com/tetafro/godot

Ok. Omw.

Comment (paologallinaharbur) [reply internal/config/config.go]: I though the same, but it is a multiline comment
